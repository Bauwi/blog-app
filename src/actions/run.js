import database from '../firebase/firebase';
import { db } from '../firebase/firebase';
import * as firebase from 'firebase';

// manage CRUD of run
export const addPostToRun = post => ({
  type: 'ADD_POST_TO_RUN',
  post
});

export const startAddPostToRun = post => (dispatch, getState) => {
  const { uid } = getState().auth;
  db
    .collection('runs')
    .doc(uid)
    .collection('UserRuns')
    .add(post)
    .then((doc) => {
      dispatch(addPostToRun({
        DBid: doc.id,
        ...post
      }));
    });
};

export const removePostToRun = id => ({
  type: 'REMOVE_POST_TO_RUN',
  id
});

export const startRemovePostToRun = (id, DBid) => (dispatch, getState) => {
  const { uid } = getState().auth;
  return db
    .collection('runs')
    .doc(uid)
    .collection('UserRuns')
    .doc(`${DBid}`)
    .delete()
    .then(dispatch(removePostToRun(id)));
};

// set Run posts. This action is dispatched in the header component
// as it needs to be available for the run dropdown

export const runsHasErrored = bool => ({
  type: 'RUN_HAS_ERRORED',
  hasErrored: bool
});

export const runIsLoading = bool => ({
  type: 'RUN_IS_LOADING',
  isLoading: bool
});

export const setRunPosts = posts => ({
  type: 'SET_RUN_POSTS',
  posts
});

export const startSetRunPosts = resume => (dispatch, getState) => {
  dispatch(runIsLoading(true));
  const { uid } = getState().auth;
  const run = [];
  return db
    .collection('runs')
    .doc(uid)
    .collection('UserRuns')
    .get()
    .then((querySnapshot) => {
      querySnapshot.forEach((doc) => {
        run.push({
          DBid: doc.id,
          ...doc.data()
        });
      });
      return run;
    })
    .then(run => dispatch(setRunPosts(run)))
    .then(() => {
      const firstUnreadIndex = run.findIndex(post => post.state === 'unread');
      if (firstUnreadIndex === -1) {
        return dispatch(setCurrentPostRun(run[0].content.id));
      }
      return dispatch(setCurrentPostRun(run[firstUnreadIndex].content.id));
    })
    .then(() => dispatch(runIsLoading(false)))
    .catch(() => dispatch(runsHasErrored(true)));
};

// change the status of the past from read to unread
export const updateRunPostToAlreadyRead = id => ({
  type: 'UPDATE_RUN_POST_TO_ALREADY_READ',
  id
});

export const startUpdateRunPostToAlreadyRead = (id, DBid) => (dispatch, getState) => {
  const { uid } = getState().auth;
  return db
    .collection('runs')
    .doc(uid)
    .collection('UserRuns')
    .doc(`${DBid}`)
    .update({ state: 'read' })
    .then(() => {
      dispatch(updateRunPostToAlreadyRead(id));
    });
};

// this one is not used anymore. But it could be used to mark some popst as unread yet
export const updateRunPostToNotAlreadyRead = id => ({
  type: 'UPDATE_RUN_POST_TO_NOT_ALREADY_READ',
  id
});

// Set which post should actually be current. Run navigation depends on this.
// instead of routing, component is updated with currentPost data.
// easier way to deal with "Resume" to provide the user the first unread post in its run.
export const setCurrentPostRun = id => ({
  type: 'SET_CURRENT_POST_RUN',
  id
});

export const startSetCurrentPostRun = id => (dispatch) => {
  dispatch(runIsLoading(true));
  return new Promise((resolve) => {
    setTimeout(() => {
      resolve();
    }, 10);
  })
    .then(() => dispatch(setCurrentPostRun(id)))
    .then(() => dispatch(runIsLoading(false)))
    .catch(() => dispatch(runsHasErrored(true)));
};

export const resetRun = () => ({
  type: 'RESET_RUN'
});

// remove all posts in tun
export const startResetRun = () => (dispatch, getState) => {
  const { uid } = getState().auth;
  return db
    .collection('runs')
    .doc(uid)
    .collection('UserRuns')
    .get()
    .then((querySnapshot) => {
      querySnapshot.forEach((doc) => {
        db
          .collection('runs')
          .doc(uid)
          .collection('UserRuns')
          .doc(doc.ref.id)
          .delete();
      });
    })
    .then(() => dispatch(resetRun()));
};

// Remove already read posts in run
export const cleanRun = () => ({
  type: 'CLEAN_RUN'
});

export const startCleanRun = alreadyRead => (dispatch, getState) => {
  const { uid } = getState().auth;
  return db
    .collection('runs')
    .doc(uid)
    .collection('UserRuns')
    .get()
    .then((querySnapShot) => {
      querySnapShot.forEach((doc) => {
        if (alreadyRead.includes(doc.ref.id)) {
          db
            .collection('runs')
            .doc(uid)
            .collection('UserRuns')
            .doc(doc.ref.id)
            .delete();
        }
      });
    })
    .then(dispatch(cleanRun()));
};
