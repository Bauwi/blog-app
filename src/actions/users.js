import { db } from '../firebase/firebase';

export const updateUser = (id, updates) => ({
  type: 'UPDATE_USER',
  id,
  updates
});

export const startUpdateUser = (id, updates) => (dispatch, getState) => {
  const uid = id || getState().auth.uid;
  return db
    .collection('users')
    .doc(uid)
    .get()
    .then((doc) => {
      if (doc.exists) {
        return db
          .collection('users')
          .doc(uid)
          .update(updates)
          .then(() => {
            dispatch(updateUser(uid, updates));
          });
      }
      return db
        .collection('users')
        .doc(uid)
        .set({
          avatar: 'http://www.perlworkshop.nl/nlpw2016/img/default-avatar.png',
          email: '',
          username: '',
          stars: 0,
          topCategories: ['life', 'programming', 'Music']
        })
        .then(startUpdateUser(uid, updates));
    });
};

export const addUserStar = nextStars => ({
  type: 'ADD_USER_STAR',
  nextStars
});

export const startAddUserStar = (id, prevStars) => dispatch =>
  db
    .collection('users')
    .doc(id)
    .update({ stars: prevStars + 1 })
    .then(() => dispatch(addUserStar(prevStars + 1)));

export const setAuthor = author => ({
  type: 'SET_AUTHOR',
  author
});

export const startSetAuthor = id => dispatch =>
  db
    .collection('users')
    .doc(id)
    .get()
    .then((doc) => {
      dispatch(setAuthor(doc.data()));
    });

export const setUserPreferences = preferences => ({
  type: 'SET_USER_PREFERENCES',
  preferences
});

export const startSetUserPreferences = preferences => (dispatch, getState) =>
  db
    .collection('users')
    .doc(getState().auth.uid)
    .get()
    .then((doc) => {
      dispatch(setUserPreferences(doc.data()));
    });
