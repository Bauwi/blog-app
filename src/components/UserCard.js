import React, { Component } from 'react';
import { Link } from 'react-router-dom';
import moment from 'moment';

class UserCard extends Component {
  static defaultProps = {
    author: {
      username: 'Anonymous',
      avatar:
        'https://www.timeshighereducation.com/sites/default/files/byline_photos/default-avatar.png',
      stars: 0,
      description:
        'I had three chairs in my house; one for solitude, two for friendship, three for society.'
    }
  };

  render() {
    const { createdAt, readingTime } = this.props.post ? this.props.post : '';
    const authorId = this.props.post ? this.props.post.authorId : this.props.userId;
    const { username, avatar, stars, description } = this.props.author;
    return (
      <div className="usercard__container">
        <div className="usercard__avatar">
          <img className="image image--circle" src={avatar} alt={`${username}'s avatar`} />
        </div>
        <div className="usercard__infos">
          <div className="usercard__infos--data">
            <Link to={`/${authorId}/read/`}>
              <h4 className="usercard__infos--data--author">{username}</h4>
            </Link>
            <p>
              {' '}
              {stars} <i className="fa fa-star" />
            </p>
          </div>
          <div className="usercard__infos--description">
            <p>{description}</p>
          </div>
          {this.props.post && (
            <footer className="usercard__infos--footer">
              <p>{moment(createdAt).format('MMM Do, YYYY')}</p>
              <p>{readingTime} minutes</p>
            </footer>
          )}
        </div>
      </div>
    );
  }
}

export default UserCard;
