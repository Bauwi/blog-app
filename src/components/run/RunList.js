import React, { Component } from 'react';
import PropTypes from 'prop-types';

import RunListItem from './RunListItem';

export default class RunList extends Component {
  renderList() {
    return this.props.posts.map((post) => {
      const key = post.content ? post.content.id : post.id;

      return <RunListItem key={key} post={post} state={post.state} />;
    });
  }

  render() {
    return (
      <div className="content-container run-list">
        {this.props.posts && this.props.posts.length !== 0 ? (
          this.renderList()
        ) : (
          <div className="run-list__empty">
            <p>Add a post to your run to start a session</p>
          </div>
        )}
      </div>
    );
  }
}

RunList.propTypes = {
  posts: PropTypes.arrayOf(PropTypes.object)
};

RunList.defaultProps = {
  posts: []
};
