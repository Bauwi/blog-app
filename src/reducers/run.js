const runReducerDefaultState = {
  isLoading: true, posts: [], hasErrored: false, current: {}
};

const runReducer = (state = runReducerDefaultState, action) => {
  switch (action.type) {
    case 'ADD_POST_TO_RUN':
      return {
        ...state,
        posts: [...state.posts, action.post]
      };
    case 'REMOVE_POST_TO_RUN':
      return {
        ...state,
        posts: state.posts.filter(post => action.id !== post.content.id)
      };
    case 'SET_RUN_POSTS':
      return {
        ...state,
        posts: action.posts
      };
    case 'RUN_IS_LOADING':
      return {
        ...state,
        isLoading: action.isLoading
      };
    case 'RUN_HAS_ERRORED':
      return {
        ...state,
        hasErrored: action.hasErrored
      };
    case 'SET_CURRENT_POST_RUN':
      return {
        ...state,
        current: state.posts.find(post => post.content.id === action.id)
      };
    case 'UPDATE_RUN_POST_TO_ALREADY_READ':
      return {
        ...state,
        current: { ...state.current, state: 'read' },
        posts: state.posts.map((post) => {
          if (post.content.id === action.id) {
            return {
              ...post,
              state: 'read'
            };
          }
          return post;
        })
      };
    case 'UPDATE_RUN_POST_TO_NOT_ALREADY_READ':
      return {
        ...state,
        posts: state.posts.map((post) => {
          if (post.content.id === action.id) {
            return {
              ...post,
              state: 'unread'
            };
          }
          return post;
        })
      };
    case 'RESET_RUN':
      return {
        posts: []
      };
    case 'CLEAN_RUN':
      return {
        ...state,
        posts: state.posts.filter(post => post.state === 'unread')
      };
    default:
      return state;
  }
};

export default runReducer;
