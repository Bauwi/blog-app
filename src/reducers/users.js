const usersReducerDefaultState = [];

const usersReducer = (state = usersReducerDefaultState, action) => {
  switch (action.type) {
    case 'CREATE_USER':
      return {
        ...state,
        preferences: action.newUser
      };
    case 'UPDATE_USER_PREFERENCES':
      return {
        ...state,
        preferences: { ...state.preferences, ...action.updates }
      };
    case 'SET_AUTHOR':
      return {
        ...state,
        author: action.author
      };
    case 'ADD_USER_STAR':
      return {
        ...state,
        author: {
          ...state.author,
          stars: action.nextStars
        }
      };
    case 'SET_USER_PREFERENCES':
      return {
        ...state,
        preferences: action.preferences
      };
    case 'USERS_IS_LOADING':
      return {
        ...state,
        isLoading: action.isLoading
      };
    case 'USERS_HAS_ERRORED':
      return {
        ...state,
        hasErrored: action.hasErrored
      };
    default:
      return state;
  }
};

export default usersReducer;
