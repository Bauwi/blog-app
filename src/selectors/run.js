const getRunPosts = (posts = [], run) => run.map(read => posts.find(post => post.id === read.id));
export default getRunPosts;
