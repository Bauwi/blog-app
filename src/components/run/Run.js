import React, { Component } from 'react';
import { connect } from 'react-redux';

import InlineList from '../InlineList';
import LoadingPage from '../LoadingPage';
import { startSetRunPosts } from '../../actions/run';

export class Run extends Component {
  state = {
    loading: true
  };
  componentWillMount() {
    if (this.props.run.length === 0) {
      this.props.startSetRunPosts().then(() => {
        this.setState(() => ({ loading: false }));
      });
    } else {
      this.setState(() => ({ loading: false }));
    }
  }

  render() {
    if (this.state.loading) {
      return <LoadingPage />;
    }
    return (
      <div>
        <p>Run</p>
        <InlineList posts={this.props.run} />
      </div>
    );
  }
}

const mapDispatchToProps = dispatch => ({
  startSetRunPosts: () => dispatch(startSetRunPosts())
});

const mapStateToProps = state => ({
  run: state.run
});

export default connect(mapStateToProps, mapDispatchToProps)(Run);
