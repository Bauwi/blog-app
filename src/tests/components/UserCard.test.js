import React from 'react';
import { shallow } from 'enzyme';
import UserCard from '../../components/UserCard';
import { user2 } from '../fixtures/users';
import run from '../fixtures/run';

test('should render UserCard component properly when post not provided', () => {
  const wrapper = shallow(<UserCard author={user2} />);
  expect(wrapper).toMatchSnapshot();
});

test('should render UserCard component properly when post provided', () => {
  const wrapper = shallow(<UserCard author={user2} post={run.posts[1].content} />);
  expect(wrapper).toMatchSnapshot();
});
