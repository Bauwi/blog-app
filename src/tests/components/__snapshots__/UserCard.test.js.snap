// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render UserCard component properly when post not provided 1`] = `
<div
  className="usercard__container"
>
  <div
    className="usercard__avatar"
  >
    <img
      alt="user2's avatar"
      className="image image--circle"
      src="http://www.perlworkshop.nl/nlpw2016/img/default-avatar.png"
    />
  </div>
  <div
    className="usercard__infos"
  >
    <div
      className="usercard__infos--data"
    >
      <Link
        replace={false}
        to="/undefined/read/"
      >
        <h4
          className="usercard__infos--data--author"
        >
          user2
        </h4>
      </Link>
      <p>
         
        20
         
        <i
          className="fa fa-star"
        />
      </p>
    </div>
    <div
      className="usercard__infos--description"
    >
      <p>
        User2 description
      </p>
    </div>
  </div>
</div>
`;

exports[`should render UserCard component properly when post provided 1`] = `
<div
  className="usercard__container"
>
  <div
    className="usercard__avatar"
  >
    <img
      alt="user2's avatar"
      className="image image--circle"
      src="http://www.perlworkshop.nl/nlpw2016/img/default-avatar.png"
    />
  </div>
  <div
    className="usercard__infos"
  >
    <div
      className="usercard__infos--data"
    >
      <Link
        replace={false}
        to="/author2Id/read/"
      >
        <h4
          className="usercard__infos--data--author"
        >
          user2
        </h4>
      </Link>
      <p>
         
        20
         
        <i
          className="fa fa-star"
        />
      </p>
    </div>
    <div
      className="usercard__infos--description"
    >
      <p>
        User2 description
      </p>
    </div>
    <footer
      className="usercard__infos--footer"
    >
      <p>
        Jan 1st, 1970
      </p>
      <p>
        12
         minutes
      </p>
    </footer>
  </div>
</div>
`;
